src/data/real/dedekind_cut.lean

```diff
@@ -0,0 +1,403 @@
+⟨λ α β γ h, begin
+  intro p,
+  simp_rw mem_add_iff,
+  rintro ⟨r, s, hr, hs, rfl⟩,
+  exact ⟨r, s, hr, h hs, rfl⟩
+end⟩
+
+-- Step 4, construction
+instance : has_coe_t Q (dedekind_cut Q) :=
+⟨λ p, { carrier := set_of (< p),
+   nonempty' := ⟨p - 1, by simp⟩,
+   ne_top' := begin
+     have : (p : Q) ∉ set_of (< (p : Q)) := by simp,
+     contrapose! this,
+     simp [this]
+   end,
+   is_lt_lower' := λ _ _, lt_trans,
+   exists_gt' := λ q hq, begin
+    simp only [set.mem_set_of_eq] at hq ⊢,
+    refine ⟨(p + q) / 2, _, _⟩,
+    { rwa [div_lt_iff, mul_two, add_lt_add_iff_left],
+      exact zero_lt_two },
+    { rwa [lt_div_iff, mul_two, add_lt_add_iff_right],
+      exact zero_lt_two }
+   end }⟩
+
+@[simp] lemma mem_coe_iff : p ∈ (q : dedekind_cut Q) ↔ p < q := iff.rfl
+
+instance : has_zero (dedekind_cut Q) := ⟨(0 : Q)⟩
+
+instance : inhabited (dedekind_cut Q) := ⟨0⟩
+
+@[simp] lemma mem_zero_iff : p ∈ (0 : dedekind_cut Q) ↔ p < 0 := iff.rfl
+
+-- Step 4 (A4)
+protected lemma add_zero (α : dedekind_cut Q) : α + 0 = α :=
+begin
+  refine le_antisymm (λ p hp, _) (λ p hp, _),
+  { simp only [mem_add_iff, mem_zero_iff, exists_and_distrib_left] at hp,
+    obtain ⟨r, hr, s, hs, rfl⟩ := hp,
+    exact α.is_lt_lower (add_lt_iff_neg_left.mpr hs) hr },
+  { obtain ⟨r, hr, hpr⟩ := α.exists_gt hp,
+    simp only [mem_add_iff, mem_zero_iff, exists_and_distrib_left],
+    exact ⟨r, hr, p - r, sub_neg.mpr hpr, add_sub_cancel'_right _ _⟩ }
+end
+
+-- Step 4 (A5) construction
+instance : has_neg (dedekind_cut Q) :=
+⟨λ α, { carrier := {p | ∃ (r : Q) (hr : 0 < r), - p - r ∉ α},
+  nonempty' := begin
+    obtain ⟨s, hs⟩ := α.exists_not_mem,
+    refine ⟨- s - 1, _⟩,
+    simp only [exists_prop, set.mem_set_of_eq, neg_sub, sub_neg_eq_add],
+    exact ⟨1, by simpa using hs⟩
+  end,
+  ne_top' := begin
+    obtain ⟨q, hq⟩ := α.nonempty,
+    suffices : - q ∉ {p : Q | ∃ (r : Q) (hr : 0 < r), -p - r ∉ α},
+    { contrapose! this,
+      simp only [this, set.top_eq_univ] },
+    rintro ⟨r, hr, hr'⟩,
+    refine hr' (α.is_lt_lower _ hq),
+    simpa using hr
+  end,
+  is_lt_lower' := begin
+    rintro p q h ⟨r, hr, hp⟩,
+    refine ⟨r, hr, λ H, hp (α.is_lt_lower _ H)⟩,
+    simpa using h
+  end,
+  exists_gt' := begin
+    rintro p ⟨r, hr, hr'⟩,
+    refine ⟨p + r / 2, ⟨r / 2, _, _⟩, _⟩,
+    { simp [div_pos_iff, hr] },
+    { rwa [neg_add, add_sub_assoc, ←sub_eq_neg_add, ←neg_add', ←mul_two, div_mul_cancel, ←neg_add',
+           add_comm, neg_add'],
+      exact zero_lt_two.ne' },
+    { simp [div_pos_iff, hr] }
+  end }⟩
+
+@[simp] lemma mem_neg_iff : p ∈ - α ↔ ∃ (r : Q) (hr : 0 < r), - p - r ∉ α := iff.rfl
+
+lemma mem_neg_imp_neg_not_mem (hp : p ∈ - α) : - p ∉ α :=
+begin
+  obtain ⟨r, rpos, hr⟩ := hp,
+  contrapose! hr,
+  refine α.is_lt_lower _ hr,
+  simpa using rpos
+end
+
+end
+
+variables [archimedean Q]
+
+-- Step 4 (A5)
+protected lemma add_left_neg (α : dedekind_cut Q) : - α + α = 0 :=
```

Comment:
It would be best to inline this into the instance below, so that way, you don't end up with an extra unused lemma.